bibcheck/plugins/geocode_all_the_things.py

```diff
@@ -0,0 +1,97 @@
+#
+# INSPIRE is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Add latitude and longitude to all records that should have them.
+
+Some conference and institution records have no latitude or longitude. This
+plugin fixes this data issue by assigning a latitude/longitude pair using the
+Google Maps geocoding API.
+
+Some records instead have only one of the two, due to a bug introduced in
+commit 9202615. This plugin also fixes this issue using the same solution.
+"""
+
+import os
+
+import googlemaps
+
+from invenio.bibrecord import record_get_field_value, record_get_field_values
+from invenio.config import CFG_GOOGLE_MAPS_API_KEY
+
+
+GMAPS = googlemaps.Client(key=CFG_GOOGLE_MAPS_API_KEY)
+
+
+def geocode(record):
+    """Uses the Google Maps API to geocode a record."""
+    def _get_approximate_address(record):
+        def _is_a_conference(record):
+            return record_get_field_values(record, '111', code='c')
+
+        if _is_a_conference(record):
+            return record_get_field_value(
+                record, '111', code='c').split(', ')
+        else:
+            city = record_get_field_value(record, '371', code='b')
+            zipcode = record_get_field_value(record, '371', code='e')
+            country = record_get_field_value(record, '371', code='d')
+            addresses = record_get_field_values(record, '371', code='a')
+
+            result = [el for el in a_values if zipcode not in el]
+            result.extend([city, zipcode, country])
+            result = [el for el in addresses if el]
+
+            return result
+
+    approximate_address = _get_approximate_address(record)
+    while approximate_address:
+        response = GMAPS.geocode(', '.join(approximate_address))
+        if response:
+            location = response[0]['geometry']['location']
+            return location['lat'], location['lng']
+
+        approximate_address = approximate_address[1:]
+
+    return None, None
+
+
+def check_record(record):
+    def _has_no_latitude_or_longitude(record):
+        return (
+            len(list(record.iterfield('034__f'))) == 0 or
+            len(list(record.iterfield('034__d'))) == 0)
+
+    def _should_have_them(record):
+        return (
+            len(list(record.iterfield('111__c'))) or
+            len(list(record.iterfield('371__d'))))
+
+    def _add_latitude_and_longitude(record, latitude, longitude):
+        subfields = [('f', str(latitude)), ('d', str(longitude))]
+        record.add_field('034__', '', subfields=subfields)
+
+    def _remove_all_latitudes_and_longitudes(record):
+        del record['034__']
```

Comment:
the record dict is keyed by MARC tag without subfield, so I think this must be `del record['034']`

```
In [1]: from invenio.bibcheck_task import AmendableRecord

In [2]: from invenio.bibedit_utils import get_bibrecord

In [3]: rec = AmendableRecord(get_bibrecord(902626))

In [4]: rec['034']
[([('d', '10.202053'), ('f', '56.1632189')], ' ', ' ', '', 22)]
```